### PR TITLE
Take node into consideration when generating hash for static pod

### LINKED ISSUE #2931: kube-proxy show up as single pod
The kube-proxy shows up as a single pod
screenshot from 2017-11-11 08 34 10
even though there is one running on each host
```
# kubectl get pods -n kube-system
NAME                                                  READY     STATUS    RESTARTS   AGE
...
kube-proxy-gke-sock-shop-default-pool-1cc5ad4b-6cf5   1/1       Running   0          1d
kube-proxy-gke-sock-shop-default-pool-1cc5ad4b-d4m6   1/1       Running   0          1d
kube-proxy-gke-sock-shop-default-pool-1cc5ad4b-wvh0   1/1       Running   0          1d
...
```
On closer inspection, this issue also shows for etcd, kube-scheduler, kube-api-server, kube-controller-manager.

The underlying problem is that we [identify certain system pods by their `kubernetes.io/config.hash` annotation](https://github.com/weaveworks/scope/blob/v1.6.5/probe/kubernetes/pod.go#L47), which turns out to be the same on all hosts, and hence scope will merge all the information into a single node (hence we see multiple kube-proxy containers and processes in the above).

We use that value instead of the usual `metadata.uid` field because for these system pods the `io.kubernetes.pod.uid` container label we use for determining which pod a container belongs to, contains that value. See #1412 for when this special case was introduced.

### PR DESCRIPTION
**Pull request by xiangpengzhao**
**What this PR does / why we need it**:
Static pods with the same spec on different nodes may have the same config.hash.
For more details, see: https://github.com/weaveworks/scope/issues/2931
Forwarded by @luxas . This also hits kubeadm.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes https://github.com/weaveworks/scope/issues/2931

**Special notes for your reviewer**:
/cc @luxas @rade @kubernetes/sig-node-pr-reviews 

**Release note**:

```release-note
action required: WARNING: static pods hash will change after the kubelet in-place upgrading. This will cause these pods restart.
```

### CODE REVIEW

**Review by bboreham — APPROVED** [review]
Putting everything in one struct is a much clearer way to generate the hash. 👍

**Comment by yujuhong** [comment]
I had an old PR #43420 fixing the same issue, but it never got in. Part of the reason was that this would cause all pods to be restarted after kubelet in-place upgrade. This change would need a release note. 
I think @tallclair also mentioned that maybe it's good that we don't allow two manifest files with the same content to exist at the same time (can't recall completely so could be inaccurate). 

Adding the original reviewers
/cc @dchen1107 @tallclair @Random-Liu

**Comment by xiangpengzhao** [comment]
>Part of the reason was that this would cause all pods to be restarted after kubelet in-place upgrade.

I didn't dig into the whole process flow of static pod. Does this sentence mean that when kubelet finds the hash of a static pod changed then it will restart the pod?

>This change would need a release note.

Relase note updated. Is there any more info we should give to users? I wonder if there is a workaround for users to make their static pods not restart in this case.

>I think @tallclair also mentioned that maybe it's good that we don't allow two manifest files with the same content to exist at the same time (can't recall completely so could be inaccurate).

Does `two manifest files with the same content to exist at the same time` mean on a single node or through the whole cluster? If it's a single node, it makes sense. But we should allow this case through the whole cluster anyway.

**Comment by errordeveloper** [comment]
@dchen1107 @tallclair @Random-Liu @yujuhong I am looking at this, and really not sure why there is a concern about pod restarts on kubelet upgrade? Are you saying that it will restart when upgrade with this change gets in place? If so, that doesn't seem like a very bad side-effect, as actually UUIDs weren't quite unique up until now, and we shouldn't have to preserve compatibility for such kind of buggy behaviour, don't you think?

**Comment by tallclair** [comment]
> Does two manifest files with the same content to exist at the same time mean on a single node or through the whole cluster?

Single node. I'd say just remove the source from the static pod object.

> I am looking at this, and really not sure why there is a concern about pod restarts on kubelet upgrade? Are you saying that it will restart when upgrade with this change gets in place? If so, that doesn't seem like a very bad side-effect, as actually UUIDs weren't quite unique up until now, and we shouldn't have to preserve compatibility for such kind of buggy behaviour, don't you think?

It's a disruptive change that breaks in-place upgrades. There are 2 possible approaches to rolling this out:
1. Use both hashes (for at least a couple of releases). New pods get the new hash value, but when reconciling pods check for the old hash as well (set it in a 'legacy hash' annotation).
2. Use a staged rollout, e.g. introduce a flag (default disabled) in 1.10 to use the new behavior, default enabled in 1.12, removed in 1.13, and highlight the action required in the release notes.

**Comment by xiangpengzhao** [comment]
>Use both hashes (for at least a couple of releases). New pods get the new hash value, but when reconciling pods check for the old hash as well (set it in a 'legacy hash' annotation).

@tallclair I implement this approach. PTAL. Thanks!

File: pkg/kubelet/config/common.go
Comment (tallclair) [inline pkg/kubelet/config/common.go]: I think this should go after generating the hash, in case it changes it.

File: pkg/kubelet/config/common.go
Comment (tallclair) [inline pkg/kubelet/config/common.go]: This block is a no-op, you can remove it.

File: pkg/kubelet/pod/pod_manager.go
Comment (tallclair) [inline pkg/kubelet/pod/pod_manager.go]: Please add a test case for this.

File: pkg/kubelet/config/common.go
Comment (yujuhong) [inline pkg/kubelet/config/common.go]: Please explain why this is needed and when it's safe to remove the code.

**Review by yujuhong — COMMENTED** [review]
Hmm...I don't think the code works as intended.

The upgraded kubelet would generate new UIDs for existing static pods with this PR. However, the running containers are labeled with old UIDs. The UID mismatch would cause kubelet to kill and restart all static pods. 
The handling of legacy hash would only help reconciling mirror pods, but re-creating mirror pods are quite harmless so this is unnecessary.

To actually make this PR work, kubelet'd have to reconcile pod UIDs while syncing the static pods -- this could be messy. I'm not convinced that it's worth adding more complexity to the codebase for ensuring no restarts for kubelet in-place upgrade (yet).

**Review by yujuhong — CHANGES_REQUESTED** [review]
Marking "request changes" to ensure my previous comment is addressed before merging :)

File: pkg/kubelet/pod/pod_manager.go
Comment (xiangpengzhao) [reply pkg/kubelet/pod/pod_manager.go]: will address your comments after @yujuhong comment is addressed :)

File: pkg/kubelet/config/common.go
Comment (xiangpengzhao) [reply pkg/kubelet/config/common.go]: Seems like this doesn't actually help to fix the issue as per your https://github.com/kubernetes/kubernetes/pull/57135#pullrequestreview-99907101. And it's never safe to remove the code as long as there are "legacy" pods and we don't want them to restart when upgrading kubelet...

/hold
until we have a final solution.

**Comment by nikopen** [comment]
interesting, this has been v1.13 for a month now but popped up only today in my radar.

@xiangpengzhao we're in code freeze, this doesn't look like it will make it for v1.13. moving it to v1.14.

I personally agree with the comment by @roffe .

/milestone v1.14

**Comment by roffe** [comment]
@xiangpengzhao @nikopen @dims

**Comment by dims** [comment]
@roffe please pull the commit into a fresh PR

**Pull request closed without merging**
